### PR TITLE
[Torch Classifier] Abstract class implementation + BERT Classifier

### PR DESCRIPTION
**Pull request by emilydinan**
- Abstract TorchClassifierAgent (only slightly modified from @samhumeau 's original internal version) meant to take away some of the overhead for training and evaluating classification models.
- Example implementation of a BERT-based classifier.

### CODE REVIEW

File: parlai/core/torch_classifier_agent.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        # class arguments
+        parser.add_argument('--classes', type=str, nargs='*', default=None,
+                            help='the name of the classes.')
+        parser.add_argument('--class-weights', type=float, nargs='*', default=None,
+                            help='weight of each of the classes for the softmax')
+        parser.add_argument('--ref-class', type=str, default=None, hidden=True,
+                            help='the class that will be used to compute '
+                                 'precision and recall. By default the first '
+                                 'class.')
+        parser.add_argument('--threshold', type=float, default=0.5,
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: I'm tempted to just say we should use argmax() but I understand why you need to tune this.

File: parlai/core/torch_classifier_agent.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+                            help='during evaluation, threshold for choosing '
+                                 'ref class; only applies to binary '
+                                 'classification')
+        # interactive mode
+        parser.add_argument('--interactive-mode', type='bool', default=False)
+        parser.add_argument('--print-scores', type='bool', default=False,
+                            help='print probability of chosen class during '
+                                 'interactive mode')
+        # miscellaneous arguments
+        parser.add_argument('--data-parallel', type="bool", default=False,
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: I don't think we want data parallel at all. The reason for data parallel in ranker agent is the ability for negative samples. In a classifier agent, where the classes are fixed, this advantage is moot and distributed will be better suited in nearly every way.

File: parlai/core/torch_classifier_agent.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        parser.add_argument('--ref-class', type=str, default=None, hidden=True,
+                            help='the class that will be used to compute '
+                                 'precision and recall. By default the first '
+                                 'class.')
+        parser.add_argument('--threshold', type=float, default=0.5,
+                            help='during evaluation, threshold for choosing '
+                                 'ref class; only applies to binary '
+                                 'classification')
+        # interactive mode
+        parser.add_argument('--interactive-mode', type='bool', default=False)
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: hmm, we can't infer this from whether a label is provided?

File: parlai/core/torch_classifier_agent.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        parser.add_argument('--print-scores', type='bool', default=False,
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: do we have a debug or verbose option elsewhere we can use instead of this specialized one?

File: parlai/core/torch_classifier_agent.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        parser.add_argument('--get-all-metrics', type='bool', default=True,
+                            help='give prec/recall metrics for all classes')
+
+    def __init__(self, opt, shared=None):
+        init_model, _ = self._get_init_model(opt, shared)
+        super().__init__(opt, shared)
+
+        # set up classes
+        if opt.get('classes') is None:
+            raise RuntimeError(
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: good

File: parlai/core/torch_classifier_agent.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+            self.reset_metrics()
+        else:
+            self.class_list = shared['class_list']
+            self.class_dict = shared['class_dict']
+            self.class_weights = shared['class_weights']
+
+        # get reference class; if opt['get_all_metrics'] is False, this is
+        # used to compute metrics
+        # in binary classfication, opt['threshold'] applies to ref class
+        if (opt['ref_class'] is None or opt['ref_class'] not in
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: no wrap needed

File: parlai/core/torch_classifier_agent.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+            raise e
+        labels_tensor = torch.LongTensor(labels_indices_list)
+        if self.use_cuda:
+            labels_tensor = labels_tensor.cuda()
+        return labels_tensor
+
+    def _update_confusion_matrix(self, batch, predictions):
+        """Update the confusion matrix given the batch and predictions.
+
+        :param batch:         a Batch object (defined in torch_agent.py)
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: docstring format

File: parlai/core/torch_classifier_agent.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+        # set up threshold, only used in binary classification
+        if len(self.class_list) == 2 and opt.get('threshold', 0.5) != 0.5:
+            self.threshold = opt['threshold']
+        else:
+            self.threshold = None
+
+        # set up model and optimizers
+        weight_tensor = torch.FloatTensor(self.class_weights)
+        self.classifier_loss = torch.nn.CrossEntropyLoss(weight_tensor)
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: add `reduce='mean'`?

File: parlai/core/torch_classifier_agent.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+
+        if self.opt.get('interactive_mode', False):
+            if self.opt.get('print_scores', False):
+                probabilities = []
+                for score in scores:
+                    probabilities.append(probs[0])
+                preds = []
+                for i, pred_id in enumerate(prediction_id):
+                    prob = round_sigfigs(probabilities[i][pred_id], 4)
+                    preds.append('{}: {}'.format(self.class_list[pred_id],
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: I think this is strange behavior. Do you not want to give it back in a structured form?

File: parlai/core/torch_classifier_agent.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+
+        for k, v in m.items():
+            m[k] = round_sigfigs(v, 4)
+
+        return m
+
+    def score(self, batch):
+        """Given a batch and labels, returns the scores.
+
+        :param batch:   a Batch object (defined in torch_agent.py)
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: docstring format

File: parlai/core/torch_classifier_agent.py (hunk 1 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: we typically just use the max, but we found this threshold with one of the projects we are using this for

File: parlai/core/torch_classifier_agent.py (hunk 3 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: i suppose, the reason i made it explicit is that we've talked about instances where we might not want to provide a label (but it's not interactive mode) but AFAIK this doesn't exist yet

File: parlai/core/torch_classifier_agent.py (hunk 6 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: what do you mean by wrap

File: parlai/core/torch_classifier_agent.py (hunk 8 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: isn't that the default?

File: parlai/core/torch_classifier_agent.py (hunk 9 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: i'll clean this up

File: parlai/core/torch_classifier_agent.py (hunk 2 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: what's the harm in keeping it?

File: parlai/core/torch_classifier_agent.py (hunk 4 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: this is only in interactive mode

File: parlai/core/torch_classifier_agent.py (hunk 3 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: i got rid of this flag in the latest updates

File: parlai/core/torch_classifier_agent.py [hunk 11]
```diff
@@ -272,9 +276,9 @@ def report(self):
     def score(self, batch):
         """Given a batch and labels, returns the scores.
 
-        :param batch:   a Batch object (defined in torch_agent.py)
-        :return:        a [bsz, num_classes] FloatTensor containing the score
-                        of each class.
+        :param Batch batch: a Batch object (defined in torch_agent.py)
+        :return: a [bsz, num_classes] FloatTensor containing the score of each
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py]: nit on the formatting but *shrug*
```
:return:
    a [bsz, num_classes] ...`
```

Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: lol i'll fix my apologies
